transport/http/outbound.go

```diff
@@ -41,61 +45,61 @@ var (
 	errOutboundNotStarted     = errors.ErrOutboundNotStarted("http.Outbound")
 )
 
-type outboundConfig struct {
-	keepAlive time.Duration
-}
+// NewOutbound is deprecated
+// TODO rename this func NewOutboundFromURL
+func NewOutbound(urlStr string) transport.Outbound {
+	agent := NewDefaultAgent()
 
-var defaultConfig = outboundConfig{keepAlive: 30 * time.Second}
+	scheme, hp, path := parseURL(urlStr)
 
-// OutboundOption customizes the behavior of an HTTP outbound.
-type OutboundOption func(*outboundConfig)
+	peerID := hostport.NewPeerIdentifier(hp)
+	peerList := peerlist.NewSingle(peerID, agent)
 
-// KeepAlive specifies the keep-alive period for the network connection. If
-// zero, keep-alives are disabled.
-//
-// Defaults to 30 seconds.
-func KeepAlive(t time.Duration) OutboundOption {
-	return func(c *outboundConfig) {
-		c.keepAlive = t
-	}
+	return NewPeerListOutbound(peerList, path, scheme)
 }
 
-// NewOutbound builds a new HTTP outbound that sends requests to the given
-// URL.
-func NewOutbound(url string, opts ...OutboundOption) transport.Outbound {
-	cfg := defaultConfig
-	for _, o := range opts {
-		o(&cfg)
+func parseURL(urlStr string) (string, string, string) {
+	parseURL, err := url.Parse(urlStr)
+
+	if err != nil {
+		return "", urlStr, ""
 	}
 
-	// Instead of using a global client for all outbounds, we use an HTTP
-	// client per outbound if unspecified.
-	client := buildClient(&cfg)
+	return parseURL.Scheme, parseURL.Host, parseURL.Path
```

Comment:
You are losing a bunch of information here. The query string (if present) is
part of the URL we want to make the request to but it's not included in
`Path`. You probably want to use `RequestURI`: https://play.golang.org/p/76xF4pvbMq

Optionally, you can just hold onto a copy of the `*URL` on the outbound rather
than the different parts of the URL. `createRequest` can then copy the `URL`,
change the `Host` and build a `http.Request` from the already-parsed URL
rather than concatenating the URL and parsing it again.